app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt

```diff
@@ -954,114 +992,701 @@ class ExplorationActivityTest {
   // TODO(#89): The ExplorationActivity takes time to finish. This test case is failing currently.
   @Test
   @Ignore("The ExplorationActivity takes time to finish, needs to fixed in #89.")
-  fun testExpActivity_onBack_showsStopExpDialog_leave_closesExpActivity() {
+  fun testExpActivity_showUnsavedExpDialog_leave_closesExpActivity() {
+    setUpAudioForFractionLesson()
     explorationActivityTestRule.launchActivity(
       createExplorationActivityIntent(
         internalProfileId,
         FRACTIONS_TOPIC_ID,
         FRACTIONS_STORY_ID_0,
-        FRACTIONS_EXPLORATION_ID_0
+        FRACTIONS_EXPLORATION_ID_0,
+        isCheckpointingEnabled = false
       )
     )
+    explorationDataController.startPlayingExploration(
+      internalProfileId,
+      FRACTIONS_TOPIC_ID,
+      FRACTIONS_STORY_ID_0,
+      FRACTIONS_EXPLORATION_ID_0,
+      isCheckpointingEnabled = false
+    )
+    testCoroutineDispatchers.runCurrent()
+
     pressBack()
     onView(withText(R.string.stop_exploration_dialog_leave_button)).inRoot(isDialog())
       .perform(click())
     assertThat(explorationActivityTestRule.activity.isFinishing).isTrue()
   }
 
-  private fun createExplorationActivityIntent(
-    internalProfileId: Int,
-    topicId: String,
-    storyId: String,
-    explorationId: String
-  ): Intent {
-    return ExplorationActivity.createExplorationActivityIntent(
-      ApplicationProvider.getApplicationContext(),
+  @Test
+  fun testExpActivity_showUnsavedExpDialog_cancel_checkOldestProgressIsSaved() {
+    explorationCheckpointTestHelper.saveFakeExplorationCheckpoint(internalProfileId)
+    setUpAudioForFractionLesson()
+    explorationActivityTestRule.launchActivity(
+      createExplorationActivityIntent(
+        internalProfileId,
+        FRACTIONS_TOPIC_ID,
+        FRACTIONS_STORY_ID_0,
+        FRACTIONS_EXPLORATION_ID_0,
+        isCheckpointingEnabled = false
+      )
+    )
+    explorationDataController.startPlayingExploration(
       internalProfileId,
-      topicId,
-      storyId,
-      explorationId,
-      /* backflowScreen= */ null
+      FRACTIONS_TOPIC_ID,
+      FRACTIONS_STORY_ID_0,
+      FRACTIONS_EXPLORATION_ID_0,
+      isCheckpointingEnabled = false
+    )
+    testCoroutineDispatchers.runCurrent()
+
+    pressBack()
+    onView(withText(R.string.stop_exploration_dialog_cancel_button)).inRoot(isDialog())
+      .perform(click())
+
+    explorationCheckpointTestHelper.verifyExplorationProgressIsSaved(
+      internalProfileId,
+      FAKE_EXPLORATION_ID_1
     )
   }
 
-  private fun setupAudio() {
-    // Only initialize the Robolectric shadows when running on Robolectric (and use reflection since
-    // Espresso can't load Robolectric into its classpath).
-    if (isOnRobolectric()) {
-      val dataSource = createAudioDataSource(
-        explorationId = RATIOS_EXPLORATION_ID_0, audioFileName = "content-en-057j51i2es.mp3"
+  // TODO(#89): Check this test case too.
```

Comment:
can we update the comment in a bit more understandable way?

Reply:
done